sql/viewregistry.go

```diff
@@ -0,0 +1,128 @@
+package sql
+
+import (
+	"strings"
+	"sync"
+
+	"gopkg.in/src-d/go-errors.v1"
+)
+
+var (
+	ErrExistingView    = errors.NewKind("the view %s.%s already exists in the registry")
+	ErrNonExistingView = errors.NewKind("the view %s.%s does not exist in the registry")
+)
+
+// A View is defined by a Node and has a name.
+type View struct {
+	name       string
+	definition Node
+}
+
+// Creates a View with the specified name and definition.
+func NewView(name string, definition Node) View {
+	return View{name, definition}
+}
+
+// Returns the name of the view.
+func (view *View) Name() string {
+	return view.name
+}
+
+// Returns the definition of the view.
+func (view *View) Definition() Node {
+	return view.definition
+}
+
+// Views are scoped by the databases in which they were defined, so a key in
+// the view registry is a pair of names: database and view.
+type viewKey struct {
+	dbName, viewName string
+}
+
+// Creates a viewKey ensuring both names are lowercase.
+func newViewKey(databaseName, viewName string) viewKey {
+	return viewKey{strings.ToLower(databaseName), strings.ToLower(viewName)}
+}
+
+// ViewRegistry is a map of viewKey to View whose access is protected by a
+// RWMutex.
+type ViewRegistry struct {
+	mutex sync.RWMutex
+	views map[viewKey]View
+}
+
+// Creates an empty ViewRegistry.
+func NewViewRegistry() *ViewRegistry {
+	return &ViewRegistry{
+		views: make(map[viewKey]View),
+	}
+}
+
+// Adds the view specified by the pair {database, view.Name()}, returning
+// an error if there is already an element with that key.
+func (registry *ViewRegistry) Register(database string, view View) error {
```

Comment:
In Go, `r` would be preferred to `registry` in the receivers

Reply:
I have seen that this is the common case, but don't really understand why; the only case where a one-letter variable seems useful to me is in counters of really small loops. But if we are following this convention, then I'll stick to it.

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#receiver-names
https://research.swtch.com/names

Reply:
Thanks for the links, @erizocosmico 

> A name's length should not exceed its information content.

Agree, but I also think that its length should be ~ directly proportional to the length of its scope.

> familiarity admits brevity

Ok, I'll buy it.

Reply:
Done in https://github.com/src-d/go-mysql-server/pull/858/commits/685c869a3f496f49b7ae49a1c4847e57f673364d.